PWGHF/TableProducer/candidateSelectorDsToKKPi.cxx

```diff
@@ -43,47 +44,92 @@ struct HfCandidateSelectorDsToKKPi {
   Configurable<LabeledArray<double>> cuts{"cuts", {hf_cuts_ds_to_k_k_pi::cuts[0], nBinsPt, nCutVars, labelsPt, labelsCutVar}, "Ds candidate selection per pT bin"};
 
-  /// Candidate selections
+  /// Candidate selections independent from the daugther-mass hypothesis
   /// \param candidate is candidate
-  /// \param trackKaon1 is the first track with the kaon hypothesis
-  /// \param trackKaon2 is the second track with the kaon hypothesis
-  /// \param trackPion is the second track with the pion hypothesis
   /// \return true if candidate passes all cuts
-  template <typename T1, typename T2>
-  bool selection(const T1& candidate, const T2& trackKaon1, const T2& trackKaon2, const T2& trackPion)
+  template <typename T1>
+  bool selection(const T1& candidate)
   {
     auto candpT = candidate.pt();
     int pTBin = findBin(binsPt, candpT);
     if (pTBin == -1) {
       return false;
     }
-    // check that the candidate pT is within the analysis range
-    if (candpT < ptCandMin || candpT > ptCandMax) {
+
+    if (candpT < ptCandMin || candpT > ptCandMax) { // check that the candidate pT is within the analysis range
+      return false;
+    }
+    if (candidate.decayLength() < cuts->get(pTBin, "decay length")) {
+      return false;
+    }
+    if (candidate.decayLengthXYNormalised() < cuts->get(pTBin, "normalized decay length XY")) {
+      return false;
+    }
+    if (candidate.cpa() < cuts->get(pTBin, "cos pointing angle")) {
+      return false;
+    }
+    if (candidate.cpaXY() < cuts->get(pTBin, "cos pointing angle XY")) {
+      return false;
+    }
+    if (std::abs(candidate.impactParameterXY()) > cuts->get(pTBin, "impact parameter XY")) {
+      return false;
+    }
+    return true;
+  }
+
+  /// Candidate selections for the KKPi daugther-mass hypothesis
+  /// \param candidate is candidate
+  /// \param trackKaon1 is the first track with the kaon hypothesis
+  /// \param trackKaon2 is the second track with the kaon hypothesis
+  /// \param trackPion is the track with the pion hypothesis
+  /// \return true if candidate passes all cuts
+  template <typename T1, typename T2>
+  bool selectionKKPi(const T1& candidate, const T2& trackKaon1, const T2& trackKaon2, const T2& trackPion)
+  {
+    int pTBin = findBin(binsPt, candidate.pt());
+    if (pTBin == -1) {
       return false;
     }
-    // cut on daughter pT
+
     if (trackKaon1.pt() < cuts->get(pTBin, "pT K") || trackKaon2.pt() < cuts->get(pTBin, "pT K") || trackPion.pt() < cuts->get(pTBin, "pT Pi")) {
       return false;
     }
-    // invariant-mass cut
-    if (std::abs(invMassDsToKKPi(candidate) - RecoDecay::getMassPDG(pdg::Code::kDS)) > cuts->get(pTBin, "m") && (std::abs(invMassDsToPiKK(candidate) - RecoDecay::getMassPDG(pdg::Code::kDS)) > cuts->get(pTBin, "m"))) {
+    if (std::abs(invMassDsToKKPi(candidate) - RecoDecay::getMassPDG(pdg::Code::kDS)) > cuts->get(pTBin, "deltaM")) {
```

Comment:
Why did you drop the cut on `invMassDsToPiKK`?

Reply:
I separated the selections depending on the mass hypothesis into two different functions, see L116. 
Actually, the previous implementation was incorrect, since the candidate had to fail the checks for both mass hypotheses to be rejected and one could imagine (although very unlikely) that a candidate could pass the mass selection for one mass hypothesis and the rest of the selections for the other one.